VideoPlayer

FernetMenta opened the pull request:
It is impossible to separate distinct work packages from my VideoPlayer branch. Here come almost all changes
- rename dvdplayer to VideoPlayer
- refactoring for hw decoders and renderers
- VideoPlayer owns video renderer
- RenderManager is not global anymore
- audio sync was moved from player into AE
- rework a/v sync
- refactor of switching refresh rates
- refactor of audio passthrough: iex packing is optional and done one if sink requests it
- dithering for GL rendering, credits to @laurimyllari 
- multi video streams by @ace20022 
- dropped libstagefright
- add EGL for X11, zero copy rendering for vaapi

FernetMenta committed:
pthreads: use mutex protocol PTHREAD_PRIO_INHERIT

popcornmix committed:
[videoplayer] Use a copy of stream info for thread safety

The stream info data is queried from multiple threads.
This can cause an issue with omxplayer which destroys and creates
the streams if it needs to switch to videoplayer mode.

popcornmix committed:
configure: Add raspberry-pi2 platform

laurimyllari committed:
gl: split output stage from yuv2rgb shader

In preparation for new features, create a separate output stage that can
be attached either to yuv2rgb or convolution filter shader. This stage
will include things like color management, RGB range conversion and
dithering.

laurimyllari committed:
gl: split output stage from convolution shaders

laurimyllari committed:
gl: GLSLOutput helper with dithering

Implement a GLSLOutput class that can be passed to shaders to implement
an output stage with RGB range conversion and dithering. Add support for
GLSLOutput in YUV2RGBShader and ConvolutionFilterShader.

Single pass rendering to full range is not optimal as the range
conversion could be done in colorspace conversion matrix. Keeping video
levels avoids banding in multipass rendering though.

laurimyllari committed:
settings: add video dither

FernetMenta committed:
renderer: promote interface up to BaseRenderer

FernetMenta committed:
renderer: refactor RenderManager and LinuxRendererGL

FernetMenta committed:
X11: add EGL

laurimyllari committed:
gl: disable dither texture after rendering only when dithering is used

FernetMenta committed:
renderer: fix forced deinterlacing for non-render methods

FernetMenta committed:
dvdplayer: fix dropped frame counter - it counted not all drops

FernetMenta committed:
renderer: fix HasFrame

FernetMenta committed:
make video renderer non-global

FernetMenta committed:
move VideoRenderers

FernetMenta committed:
rename dvdplayer to VideoPlayer

FernetMenta committed:
AE: refactor IAudioCallback

FernetMenta committed:
move audio sync to AE

popcornmix committed:
[omxplayer] Update to support VideoPlayer renderer

popcornmix committed:
[MMAL] Update to support VideoPlayer renderer

popcornmix committed:
[GLES] Fix build with VideoPlayer

popcornmix committed:
[gles] Fix build with gles. GL_RED is not defined

notspiff committed:
fixed: don't assing std::string to nullptr

notspiff committed:
fixed: macro existence should not rely on config.h usage

notspiff committed:
fixed: unify include paths with other files

FernetMenta committed:
VideoPlayer: rework HandlePlayspeed

FernetMenta committed:
dvdplayer: you are dead, dont you come back anymore'

wsnipex committed:
[depends] fix target raspberry-pi

Memphiz committed:
[shader] - fix compilation of glsl shaders for gles

Memphiz committed:
[shader] - use GLSLOutput object on gles platforms aswell

Memphiz committed:
[ios] - add implementation for RendererVTB

Memphiz committed:
[VideoRendererGLES] - refactor to support derived hw dec renderer - similar to VideoRendererGL (atm only VTB is implemented. All other hw dec renderers are still in VideoRendererGLES and need to be moved into seperate implementations by the platform devs - similar to VideoRendererGL (atm only VTB is implemented. All other hw dec renderers are still in VideoRendererGLES and need to be moved into seperate implementations by the platform devs))

Memphiz committed:
[RenderManager] - hookup hw dec RendererVTB

Memphiz committed:
[HwRendererVTB] - add to CMakeLists.txt and Makefile buildsys

Memphiz committed:
[RendererVTB] - some more cleanup

Memphiz committed:
[HwRendererIMX] - moved IMX video rendering out of LinuxRendererGLES.cpp
into its own class

Memphiz committed:
[HwRendererOMX] - moved OpenMax (OMX) video rendering out of
LinuxRendererGLES.cpp into its own class

Memphiz committed:
[HwDecRenderers] - get some more platform specific stuff out of
RendererGLES by asking the hwdecoder for the supported render formats

Memphiz committed:
[HwDecRendererStagefrighht] - moved OpenMax (OMX) video rendering out of LinuxRendererGLES.cpp into its own class

Memphiz committed:
[HwDecRendererMediaCodec] - moved Android MediaCodec (surface) video rendering out of LinuxRendererGLES.cpp into its own class

AlwinEsch committed:
Add X11 EGL depends build support

AlwinEsch committed:
[VideoPlayer] cleanup DVDVideoCodec.*

Memphiz committed:
[VideoRenderer/GLES] - drop AddSupportedHwRenderFormats - m_formats will
go away in the future

Memphiz committed:
remove stagefright

notspiff committed:
changed: allow multiple callbacks in AE

FernetMenta committed:
RenderManager - cleanup CreateRenderer

wolfgar committed:
Fix iMX6 compilation

FernetMenta committed:
VideoPlayer: flush renderer when skin is changed

wsnipex committed:
[EGL] don't always include amlogic

Commit:
[win32] Fix compilation errors.

FernetMenta committed:
VideoPlayer: keep dvdplayer path for libs on Windows, temp

FernetMenta committed:
VideoPlayer: limit sleep time for render to 500ms

FernetMenta committed:
rename OnLostDevice, OnResetDevice

popcornmix committed:
fixup: Fix GLES and MMAL builds

FernetMenta committed:
VideoPlayer: honor flac metatags for channel layouts

philipl committed:
ffmpeg: Update AVPixelFormat and AV_PIX_FMT_* to compile with master

The deprecated PixelFormat and PIX_FMT_* names have been removed in
ffmpeg master.

FernetMenta committed:
set desired resolution before start of playback

FernetMenta committed:
VideoPlayer: clean audio interface

FernetMenta committed:
disable passthrough when sync playback to display

FernetMenta committed:
VideoPlayer: drop useless masterclock

FernetMenta committed:
X11: delayed OnResetDisplay after change of refresh rate

FernetMenta committed:
VideoPlayer: rework player sync to clock

wolfgar committed:
Fix imx buffer lock/release to have proper ref count

popcornmix committed:
[omxplayer] Update to new VideoPlayer sync scheme

fritsch committed:
Advancedsettings: Remove blackbarcolour and set it automatically

FernetMenta committed:
VideoPlayer: drop annoying asked for new data packet msg

FernetMenta committed:
vaapi: do not disassociate egl context

popcornmix committed:
[omxplayer] Ensure MMAL renderer is used in BYPASS mode

FernetMenta committed:
VideoRenderer: fix state after flush

FernetMenta committed:
VideoPlayer: allow speed up to 1.5

koying committed:
Fixup mediacodec

FernetMenta committed:
VideoPlayer: flag and handle realtime streams

FernetMenta committed:
VideoPlayer: add debug log for flushing audio

FernetMenta committed:
VideoPlayer: do not flush and drain audio on stall

koying committed:
Get rid of libstagefright

popcornmix committed:
[renderer] Remove RenderUpdateCallBack and RenderFeaturesCallBack

popcornmix committed:
[AE] Fix large buffertime check

popcornmix committed:
AE: fix calculation of bytesToSkip to avoid cracking sounds

FernetMenta committed:
VideoPlayer: fix cachestate after FlushBuffers

koying committed:
baseline amlcodec

koying committed:
fixup aml

popcornmix committed:
squash: fix wrong prototype for RenderUpdateVideoHook

FernetMenta committed:
vaapi: cope with comtinuous get_buffer errors

FernetMenta committed:
VideoPlayer: fix ff after recent changes

lkosmid committed:
Fixed compilation of refactored EGL on X11

FernetMenta committed:
merge error

FernetMenta committed:
VideoPlayer: pass CFileItem to input stream

FernetMenta committed:
vtb: rename after merge

FernetMenta committed:
AE: init sample pts

FernetMenta committed:
AE: move iec packing for passthrough into AE, contains squashed code from fritsch

popcornmix committed:
omxplayer: fixup: AE: test type of stream for passthough modes

Commit:
[win32] Add handling of DispResResource.

Commit:
[win32] Change way to delayed dll load.

fritsch committed:
AESinkALSA: Explicitely push supported Stream Types and AE_FMT_RAW

fritsch committed:
AESinkAudiotrack: Add Stream Types

fritsch committed:
AESink: Remove m_frameSamples from sink implementations

fritsch committed:
AEUtils: Add Streamtypes to Enumeration output

fritsch committed:
AEUtils: Fix DataFormatToStr with new AE_FMT_RAW

FernetMenta committed:
AE: drop old raw data formats

fritsch committed:
DVDAudioCodecPassthrough: Set samplerate when creating StreamInfo

FernetMenta committed:
AE: update OSX sink

fritsch committed:
AESinkDirectSound: Prepare for new STREAM_TYPES

fritsch committed:
AESinkWASAPI: Prepare for new STREAM_TYPES

fritsch committed:
AESinkDARWINIOS: Port to new STREAM_TYPE

popcornmix committed:
[OMXAudio] Remove obsolete hw decode support

popcornmix committed:
AE: update omx audio

popcornmix committed:
AE: update Pi Sink

popcornmix committed:
[OmxAudio] Make m_format more local

FernetMenta committed:
AE: set channel layout for iec packing

FernetMenta committed:
AE: fixes (truehd) and cleanup

FernetMenta committed:
VideoPlayer: trash method OpenStreamPlayer - enables detection if stream was actually opened or not

FernetMenta committed:
RenderSystem: reduce video latency during video playback - GL part

FernetMenta committed:
VideoPlayer: refactor resolution switching

FernetMenta committed:
VideoPlayer: reqork dithering

FernetMenta committed:
VideoPlayer: win32 fixes for ffmpeg

FernetMenta committed:
VideoPlayer: reslotion fixes

FernetMenta committed:
sync XCode project

FernetMenta commented:
@afedchin VS filters (not sure about project) had merge conflicts I was not able to resolve

koying commented:
Obviously quite difficult to review, so I'll trust you ;)
How do you want to proceed? Merge and go from there?

FernetMenta commented:
good question. it depends on the available time of the platform devs. imo it should at least compile for all targets before merge. what do you suggest?

hudokkow commented:
As it stands, this PR almost crashes the browser when you hit "Files changed" above and then becomes almost unusable...

I say if it builds on all targets and runtime is OK(ish), merge. Last time I tested your branch was over a week ago and it worked. I trust this will too.
Anyway, I'm going to pull this PR and start testing on Ubuntu tonight.

ace20022 commented:
@FernetMenta Could you please or replace my commit with https://github.com/ace20022/xbmc/commit/56ec8165232ef38482b1e87235fe895adb682cbb

There's one wrong function call in it ;)

FernetMenta commented:
@ace20022 will do

FernetMenta commented:
@popcornmix could you please verify omxplayer. due to reorder and squash of commits I see a small diff to my current master

FernetMenta commented:
jenkins build this please

afedchin commented:
@FernetMenta vs filters looks good. let's wait what jenkins will say

koying commented:
@FernetMenta Agreed that it should pass jenkins. 
I find it easier to "functionally" fix on master thereafter

popcornmix committed:
[omxplayer] Remove debug code that uses removed GetCacheTotal api

fritsch commented:
Expect PRs for the Windows Sinks - I did those blindly ... and DirectSound will fail (damn braket missing).

Wohooo was a long time.

MilhouseVH commented:
This is a debug log from latest kodi master (OpenELEC, Nvidia_Legacy) with the latest version of this PR: http://sprunge.us/WbEU

Video playback hangs at startup for almost 2 minutes, before eventually playing, but only at about 2 frames per second (ie. very jerky). This behaviour is the same with or without passthrough audio.

An identical build using your [latest master](https://github.com/xbmc/xbmc/compare/xbmc:fde0686b7845e216c74d0305b803569c6e2288f5...FernetMenta:a99657f89ac1bc3dc24514483d279a5ad905f379) in place of #8486 is working normally, so maybe some commits are missing from #8486?

hudokkow commented:
I've been playing with this PR for the last 3h and everything is OK. Tried a few movies, TV shows, Youtube videos, etc. It played everything I tried. BTW, fantastic start playing times. Great stuff!
Tested with legacy hardware: C2D T4400@2.20GHz, RV710/M92 [Mobility Radeon HD 4530/4570/545v], 4GB RAM, etc...

FernetMenta commented:
@MilhouseVH 
my master is rebased on videoplayer branch. the remaining commits are floating on top. most of them are for cmake. I can only think of those that can make a difference:
https://github.com/FernetMenta/xbmc/commit/4132b68c1965a2f9e4c9458db4eba0d836f79d74
https://github.com/FernetMenta/xbmc/commit/c21fe9a63e8c17bc6727591091024177ea5f6a2a

fritsch commented:
I sent a PR for the Wasapi build failure ... we need someone with windows that can actually test the resulting Sink ... as some structs have changed. I am not totally sure I did not miss something or set wrong values from time to time ...

MilhouseVH commented:
>for cmake. I can only think of those that can make a difference:
> FernetMenta@4132b68
> FernetMenta@c21fe9a

Yep, with both of those playback is working - let me narrow down which is required.

MilhouseVH commented:
I need the PTHREADS commit: FernetMenta@4132b68 - with just this added, playback is working.

My setup: OpenELEC master, NVidia_Legacy (Revo 3700 with ION2), Onkyo TX-NR828 AVR.

FernetMenta commented:
@Memphiz @wsnipex the Linux VM is missing libEGL. Could you add it please?

fritsch commented:
Let's see what happens: http://jenkins.kodi.tv/job/WIN-32/6881/

FernetMenta committed:
VideoPlayer: remove some libstagefright remnants

wsnipex commented:
I've tried my best to get the needed mesa libs installed on the slave. If it doesn't work, we'll have to set up a new one as we're still on ubuntu 12.04...

wsnipex commented:
jenkins build this please

FernetMenta committed:
addon ae: fixups

FernetMenta committed:
fixup VS project

ace20022 commented:
@FernetMenta please drop my commit on the next occasion

FernetMenta commented:
@ace20022 done

jenkins build this please

FernetMenta commented:
> ld: library not found for -lpython2.6

@wsnipex ^^ where does this come from?

FernetMenta commented:
@wsnipex my guess is that buildsteps do not force make xcode_depends and we toggle between Jarvis branch and master

FernetMenta committed:
VideoPlayer: fix compile error on Android

koying commented:
@FernetMenta There's another one: https://github.com/koying/xbmc/commit/99dd50a4a07b36f8640d7456b64dd749159cc99a

FernetMenta commented:
@koying thanks
@wsnipex disregard, merge issue here

FernetMenta commented:
jenkins build this please

koying committed:
videoplayer: mediacodec surface

koying commented:
@FernetMenta Please pick https://github.com/koying/xbmc/commit/01be58f64f44cd3d21c2fd08833e8201b0c53f93 for mediacodec surface

hudokkow commented:
Android X86 failed with
> In file included from platform/Implementation.cpp:22:0:
/home/jenkins/workspace/Android-X86/xbmc/threads/platform/pthreads/Implementation.cpp: In function 'bool XbmcThreads::pthreads::setRecursiveAttr()':
/home/jenkins/workspace/Android-X86/xbmc/threads/platform/pthreads/Implementation.cpp:42:54: error: 'PTHREAD_PRIO_INHERIT' was not declared in this scope
         pthread_mutexattr_setprotocol(&recursiveAttr,PTHREAD_PRIO_INHERIT);
                                                      ^
/home/jenkins/workspace/Android-X86/xbmc/threads/platform/pthreads/Implementation.cpp:42:74: error: 'pthread_mutexattr_setprotocol' was not declared in this scope
         pthread_mutexattr_setprotocol(&recursiveAttr,PTHREAD_PRIO_INHERIT);
                                                                          ^

Win32 didn't even run. new test build: http://jenkins.kodi.tv/job/WIN-32/6891/

FernetMenta committed:
OSX: remove remnants to libGLEW

koying commented:
pthread_mutexattr_setprotocol not in NDK altogether, so needs to exclude android as a whole, not just arm

FernetMenta commented:
> pthread_mutexattr_setprotocol not in NDK altogether

yes, already done but forgot to add on the last run

FernetMenta commented:
jenkins build this please

MartijnKaijser commented:
all passed :)

anaconda commented on xbmc/cores/AudioEngine/Interfaces/AEStream.h:
```diff
@@ -29,6 +29,32 @@ extern "C" {
+  {
+    SYNC_OFF,
+    SYNC_PLAY,
+    SNYC_ACTIVE
```
Typo

popcornmix commented:
I'm okay to disable it on Pi1. Just checking if an ifdef or a settings option is preferred.
We could leave the setting in place but make it enabled/not-visible by default.

FernetMenta commented:
> We could leave the setting in place but make it enabled/not-visible by default.

that is like dead code and an abuse of settings. the setting has to go

FernetMenta commented:
So finally this builds on all targets. I forced pushed after:

- fixed a typo, thanks @anaconda 
- dropped a commit about setting of dca codec, <- @popcornmix 
- fixed a/v sync of passthrough audio, @MilhouseVH 

any objections to get this in after a final build?

popcornmix commented:
No objections from me.

fritsch commented:
Also fine by me as well. It was acked by all towers and contains. Will need
further fixes for audio on Windows and work for android though.
Am 06.12.2015 9:22 nachm. schrieb "popcornmix" <notifications@github.com>:

> No objections from me.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/8486#issuecomment-162344624>.
>

FernetMenta commented:
jenkins build this please

ace20022 commented:
Just a note and without a log, platform windows: There's a picture freeze and a very high memory consumption during rw/ff 8/16/32x with this test file: http://ftp.nluug.nl/pub/graphics/blender/demo/movies/ToS/tearsofsteel_4k.mov But there's no hang, playback resumes after pressing play.

MilhouseVH commented:
I have a reproducible crash with this latest PR.

1. Enable passthrough
2. Enable "Dolby Digital (AC3) capable receiver"
3. Play a video with AC3 soundtrack
4. Seek within the video (ie. jump to 99 seconds with `9-9-ENTER`)
5. Kodi will crash a second or two after completing the seek

I can reproduce this on both x86 and RPi2 (both OpenELEC).

Crashlog from RPi2: http://sprunge.us/WiQU (debug-enabled build)
Crashlog from x86 (Nvidia_Legacy, Revo3700 w/ION2): http://sprunge.us/QfJJ (not debug-enabled build)

koying commented:
Congrats :)

Le lun 7 déc. 2015 07:38, Rainer Hochecker <notifications@github.com> a
écrit :

> Merged #8486 <https://github.com/xbmc/xbmc/pull/8486>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/8486#event-484051500>.
>

FernetMenta commented:
hehe :)
I want to mention this again here: we know that there is still work to do but it is much easier to do this on here on master instead of another branch that would diverge even more.

fritsch commented:
Yeah - that's what we can do in master now. But let's not forget on those outstanding issue for v16.

MilhouseVH commented:
Just to say the AC3 crash is fixed in latest build, possibly because of https://github.com/FernetMenta/xbmc/issues/340

FernetMenta commented:
@popcornmix friendly reminder for an alternative to https://github.com/FernetMenta/xbmc/commit/67eb6156e611d5e249a1962ce60c14d551e8b5f5
I won't forget this :)

koying commented:
I'm all to drop the setting, but we have to check whether channel layout > 2.0, or it makes no sense imo

popcornmix commented:
@FernetMenta okay, I'll have a look.
